docs/generic-interfaces-infrastructure.md

```diff
@@ -0,0 +1,426 @@
+
+Next, an interface is needed for each "Methods" implementation, for each generic interface. For `IIterator<T>`, it looks like this:
+
+```csharp
+public interface IIteratorMethods<T>
+{
+    static abstract T Current(WindowsRuntimeObjectReference objectReference);
+
+    static abstract bool HasCurrent(WindowsRuntimeObjectReference objectReference);
+
+    static abstract bool MoveNext(WindowsRuntimeObjectReference obj);
+
+    static abstract uint GetMany(WindowsRuntimeObjectReference obj, Span<T> items);
+}
+```
+
+This exactly matches the shape that any "Methods" type would have for any given projected interface.
+
+To pair each of these "Methods" interfaces, we have a "Methods" type for the mapped .NET interface type. For instance:
+
+```csharp
+public static class IEnumeratorMethods<T>
+{
+    public static T Current<TMethods>(WindowsRuntimeObjectReference objectReference)
+        where TMethods : IIteratorMethods<T>
+    {
+        return TMethods.Current(objectReference);
+    }
+
+    public static bool HasCurrent<TMethods>(WindowsRuntimeObjectReference objectReference)
+        where TMethods : IIteratorMethods<T>
+    {
+        return TMethods.HasCurrent(objectReference);
+    }
+
+    // Other methods...
+}
+```
+
+These methods are also generic on a given "Methods" interface implementation, and can implement the mapped functionality on top of the underlying Windows Runtime interface logic, where needed. For `IEnumerator<T>`, the implementation is trivial, but for other interfaces (eg. `IList<T>`) there is a fair amount of additional "semantics adjustments" that is required, that can all be done here. This also allows CsWinRT to be versioned easily in this area, as all this logic on top of the underlying APIs is just in `WinRT.Runtime.dll`.
+
+Lastly, we also need to introduce some additional building blocks to wire things up at marshalling time. First, a new interface:
+
+```csharp
+public interface IWindowsRuntimeUnsealedComWrappersCallback
+{
+    static abstract bool TryCreateObject(
+        void* value,
+        ReadOnlySpan<char> runtimeClassName,
+        [NotNullWhen(true)] out object? result);
+}
+```
+
+This is conceptually similar to `IWindowsRuntimeComWrappersCallback` (which is used for delegates and sealed runtime class names), with the main difference being that it is specialized for unsealed types (and interfaces). In this case, implementations need to know the runtime class name of the native object, to determine whether they can marshal it via the fast path using a statically visible RCW type (see below), or whether they should fallback to the logic to marshal arbitrary opaque object that is used when no static type information is available (which involves going through the interop type lookup to get the marshalling info, etc.).
```

Comment:
Pretend I don't know anything about the current C#/WinRT implementation in this regard. What do I need to know/care about to evaluate this proposal? :)